README.md
# NGA Deep Learning

This repository includes a set of general scripts to preprocess, train, and predict
land cover and land use characteristics from very high-resolution imagery in GeoTIF format.

## Business Case

The following repository performs semantic segmentation classification over GeoTIF data using
convolutional neural networks (CNN), particularly a modified UNet.

## Table of Contents

1. [ Computational Resources ](#Computational_Resources)
   * [ Conda Environment Installation ](#Conda_Environment_Installation)
   * [ Container Environment Installation ](#Container_Environment_Installation)
2. [ How-To ](#How_To)
2. [ Authors ](#Authors)
3. [ Contributors ](#Contributors)
4. [ References ](#References)
5. [ Acknowledgments ](#Acknowledgments)

## Computational Resources <a name="Computational_Resources"></a>

A system with NVIDIA GPUs is required to run the scripts located in this repository.

### Conda Environment Installation <a name="Conda_Environment_Installation"></a>

Installing conda environment.

``` bash
module load anaconda
git clone https://github.com/jordancaraballo/nga-deep-learning.git
cd nga-deep-learning; conda env create -f requirements/environment.yml;
```

Note: Using Tensorflow 2.2 at this time since NCCS GPU Cluster does not have CUDA 11.x available as of 12/16/2020.
When CUDA 11.0 is available, then intstall cudatoolkit=10.2 and tensorflow=2.4 or greater.

### Container Environment Installation <a name="Container_Environment_Installation"></a>

The following section describes how to build, pull, and use a Docker/Singularity container to execute
the code provided in this repository.

#### Building Container

This repository includes a Docker file to build the container. The base image is taken from NVIDIA
NGC container repository with the RAPIDS environment.

```bash
git clone https://github.com/jordancaraballo/nga-deep-learning.git
cd nga-deep-learning/packages/
docker build --tag nga-deeplearning:latest .
docker login
docker tag nga-deeplearning:latest nasanccs/nga-deeplearning:latest
docker push nasanccs/nga-deeplearning
```

#### Downloading Container

Docker containers can be pulled as Singularity containers to be executed on HPC environments. The
following commands allow the download of the container from DockerHub.

```bash
singularity pull docker://docker.io/nasanccs/nga-deeplearning:latest
```

#### Executing Container

For Singularity containers to have access to other paths within the HPC environment, we need to bind
directories to particular locations in the container. The container provided by this project has a built-in
conda environment that includes all software dependencies to run this project. To shell into the container
and activate this environment:

```bash
singularity shell --nv -B /att/nobackup/$user:/att/nobackup/$user,/att/gpfsfs/atrepo01/ILAB:/att/gpfsfs/atrepo01/ILAB nga-deeplearning_latest.sif
source activate rapids
```

## How-To

### Expected Input Data

The following scripts expect data in GeoTIF format. There should be a file with raw data, preferebly TOA corrected, and a file with
labels or the mask to map with. The expected data file can have anywhere from 3 to N number of channels/bands, while the mask file should
have a single channel/band with integer values. Each integer value representing a class to classify. At the end of the day, the directories
should look similar to the following example:

```bash
/att/data/data01.tif, /att/data/data02.tif, /att/data/data03.tif
/att/labels/labels01.tif, /att/labels/labels02.tif, /att/labels/labels03.tif
```

Example input shape are (5000,5000,6) for data rasters, and (5000,5000) for mask files. The information regarding the data will be stored
in a CSV file located under scripts/config/data.csv, with the following format:

```bash
data,label,ntiles_train,ntiles_val,ymin,ymax,xmin,xmax
Keelin00_20120130_data.tif,cm_Keelin00_20120130_new_2.tif,500,1000,0,50,0,50
Keelin00_20180306_data.tif,cm_Keelin00_20180306_new_2.tif,500,1000,0,50,0,50
Keelin01_20151117_data.tif,cm_Keelin01_20151117_new.tif,500,1000,0,50,0,50
```

Where:

* data: is the filename with the data values
* label: is the filename with the mask values
* ntiles_train: number of training tiles to extract from the data file
* ntiles_val: number of validation tiles to extract from the data file
* ymin,ymax,xmin,xmax: values to exclude from training dataset to include in validation tiles

### Configuration File

### Preprocess

In this section of the pipeline we take raw raster and masks, and we generate training and validation
datasets to work with. The following script will save NPZ files into local storage with the train and mask
mappings taken from the data.csv file. Parameters are taken from the Configuration File. Expect to find two
new directories on the ROOT_DIR specified on the Configuration file with an NPZ file per raster. These NPZ
files contain the dataset mappings with 'x' for data, and 'y' for the label.

if specified, data will be stanrdardized using local standardization.

```bash
python preprocessing.py
```

### Train

In this section of the pipeline we proceed to train the model. Please refer to the Configuration file for more
details on parameters required for training. The main script will: read the data files, map them into TensorFlow
datasets, initialize the UNet model, and proceed with training. A model (.h5 file) will be saved for each epoch
that improved model performance.

You might notice that the first epoch of the model takes a long time to train. This is expected since the model
is being optimized during the first epoch. Distributed training across multiple GPUs is enabled, together with
mixed precission for additional performance enhancements. Upcoming modifications will include A100 performance
enhancements and Horovod for multi-node training.

```bash
python train.py
```

### Predict

In this section we proceed to train the desired data taking one of the saved models, and using it for predictions.
Please refer to the Configuration file for more details on parameters required for predicting. Ideally, all images to
predict will be stored in the same directory ending on .tif extensions. The script will go through every file, predict
it and output both the predicted GeoTIF and the probabilities arrays.

```bash
python predict.py
```

## Authors

* Jordan Alexis Caraballo-Vega, <jordan.a.caraballo-vega@nasa.gov>

## Contributors

* Mary Aronne, <mary.aronne@nasa.gov>

## References

[1] Chollet, François; et all, Keras, (2015), GitHub repository, https://github.com/keras-team/keras. Accessed 13 February 2020.

[2] Paszke, Adam; Gross, Sam; Chintala, Soumith; Chanan, Gregory; et all, PyTorch, (2016), GitHub repository, https://github.com/pytorch/pytorch. Accessed 13 February 2020.

[3] Google Brain Team; et all, TensorFlow, (2015), GitHub repository, https://github.com/tensorflow/tensorflow. Accessed 13 February 2020.

## Acknowledgments

* Mark Carroll, NASA Center for Climate Simulation, 606.2
* Mary Aronne, NASA Center for Climate Simulation, 606.2
* Dan Duffy, NASA Center for Climate Simulation, 606.2
* Laura E. Carriere, NASA Center for Climate Simulation, 606.2


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. Model training
and configuration files are documented here as well.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## 2021-01-12

### Added

### Changed

### Data

### Model

### Results


scripts/train.py
# --------------------------------------------------------------------------
# Preprocessing and dataset creation from NGA data. This assumes you provide
# a configuration file with required parameters and files.
# --------------------------------------------------------------------------
import os                # system modifications
import sys               # system modifications
import time              # tracking time
import numpy as np       # for arrays modifications
import cupy as cp        # for arrays modifications
import tensorflow as tf  # deep learning framework

from core.unet import unet_batchnorm         # unet network to work with
from core.utils import get_training_dataset  # getting training dataset
from core.utils import get_tensorslices      # getting tensor slices
from core.utils import gen_callbacks         # generate callbacks

# tensorflow imports
# from tensorflow.keras.mixed_precision import experimental as mixed_precision
from tensorflow.keras import mixed_precision
from tensorflow.keras.optimizers import Adam
from tensorflow.keras.optimizers import Adadelta

# define configuration object
from config import Config
config = Config.Configuration()

__author__ = "Jordan A Caraballo-Vega, Science Data Processing Branch"
__email__ = "jordan.a.caraballo-vega@nasa.gov"
__status__ = "Development"

# Define some environment variables to help refining randomness.
# Note: there might still be some randomness since most of the code
# is ran on GPU and sometimes parallelization brings changes.
np.random.seed(config.SEED)
tf.random.set_seed(config.SEED)
cp.random.seed(config.SEED)

print(f"Tensorflow ver. {tf.__version__}")

# verify GPU devices are available and ready
os.environ['CUDA_VISIBLE_DEVICES'] = config.CUDA
devices = tf.config.list_physical_devices('GPU')
assert len(devices) != 0, "No GPU devices found."

# ------------------------------------------------------------------
# System Configurations
# ------------------------------------------------------------------
if config.MIRROR_STRATEGY:
    strategy = tf.distribute.MirroredStrategy()
    print('Multi-GPU enabled')

if config.MIXED_PRECISION:
    policy = mixed_precision.Policy('mixed_float16')
    mixed_precision.set_global_policy(policy)
    print('Mixed precision enabled')

if config.XLA_ACCELERATE:
    tf.config.optimizer.set_jit(True)
    print('Accelerated Linear Algebra enabled')

# Disable AutoShard, data lives in memory, use in memory options
options = tf.data.Options()
options.experimental_distribute.auto_shard_policy = \
    tf.data.experimental.AutoShardPolicy.OFF


# ---------------------------------------------------------------------------
# script train.py
# ---------------------------------------------------------------------------

def main():

    # Main function to collect configuration file and run the script
    print(f'GPU REPLICAS: {strategy.num_replicas_in_sync}')
    t0 = time.time()

    print(f'Train dir: {config.TRAIN_DATADIR}')
    print(f'Validation dir: {config.VAL_DATADIR}')

    # Initialize Callbacks
    callbacks = gen_callbacks(config, config.CALLBACKS_METADATA)

    # open files and get dataset tensor slices
    train_images, train_labels = get_tensorslices(
        data_dir=config.TRAIN_DATADIR, img_id='x', label_id='y'
    )

    # open files and get dataset tensor slices
    val_images, val_labels = get_tensorslices(
        data_dir=config.VAL_DATADIR, img_id='x', label_id='y'
    )

    # extract values for training
    NUM_TRAINING_IMAGES = train_images.shape[0]
    NUM_VALIDATION_IMAGES = val_images.shape[0]
    STEPS_PER_EPOCH = NUM_TRAINING_IMAGES // config.BATCH_SIZE

    print(f'{NUM_TRAINING_IMAGES} training images')
    print(f'{NUM_VALIDATION_IMAGES} validation images')

    # generate training dataset
    train_dataset = \
        tf.data.Dataset.from_tensor_slices((train_images, train_labels))

    # generate validation dataset
    val_dataset = tf.data.Dataset.from_tensor_slices((val_images, val_labels))
    val_dataset = val_dataset.batch(config.VAL_BATCH_SIZE)

    # Create model output directory
    os.system(f'mkdir -p {config.MODEL_SAVEDIR}')

    # Initialize and compile model
    with strategy.scope():

        # initialize UNet model
        model = unet_batchnorm(
            nclass=config.N_CLASSES, input_size=config.INPUT_SIZE,
            maps=config.MODEL_METADATA['network_maps']
        )

        # initialize optimizer, exit of not valid optimizer
        if config.MODEL_METADATA['optimizer_name'] == 'Adadelta':
            optimizer = Adadelta(lr=config.MODEL_METADATA['lr'])
        elif config.MODEL_METADATA['optimizer_name'] == 'Adam':
            optimizer = Adam(lr=config.MODEL_METADATA['lr'])
        else:
            sys.exit('Optimizer provided is not supported.')

        # enabling mixed precision to avoid underflow
        optimizer = mixed_precision.LossScaleOptimizer(optimizer)

        # compile model to start training
        model.compile(
            optimizer,
            loss=config.MODEL_METADATA['loss'],
            metrics=config.MODEL_METADATA['metrics']
        )
        model.summary()

    # Disable AutoShard, data lives in memory, use in memory options
    train_dataset = train_dataset.with_options(options)
    val_dataset = val_dataset.with_options(options)

    # Train the model and save to disk
    model.fit(
        get_training_dataset(
            train_dataset,
            config,
            do_aug=config.MODEL_METADATA['do_aug']
        ),
        initial_epoch=config.START_EPOCH,
        epochs=config.N_EPOCHS,
        steps_per_epoch=STEPS_PER_EPOCH,
        validation_data=val_dataset,
        callbacks=callbacks,
        verbose=2
    )

    print(f'Execution time: {time.time() - t0}')


# -------------------------------------------------------------------------------
# main
# -------------------------------------------------------------------------------

if __name__ == "__main__":

    main()


scripts/predict.py
# --------------------------------------------------------------------------
# Prediction from TF model and NGA data. This assumes you provide
# a configuration file with required parameters and files.
# --------------------------------------------------------------------------
import os                # system modifications
import time              # tracking time
from tqdm import tqdm    # for local progress bar
import numpy as np       # for arrays modifications
import cupy as cp        # for arrays modifications
import tensorflow as tf  # deep learning framework
import xarray as xr      # read rasters

# tensorflow imports
# from tensorflow.keras.mixed_precision import experimental as mixed_precision
from tensorflow.keras import mixed_precision
from tensorflow.keras.models import load_model

# core library imports
from core.utils import _2d_spline, arr_to_tif
from core.utils import predict_all
from core import indices

# define configuration object
from config import Config
config = Config.Configuration()

__author__ = "Jordan A Caraballo-Vega, Science Data Processing Branch"
__email__ = "jordan.a.caraballo-vega@nasa.gov"
__status__ = "Development"

# Define some environment variables to help refining randomness.
# Note: there might still be some randomness since most of the code
# is ran on GPU and sometimes parallelization brings changes.
np.random.seed(config.SEED)
tf.random.set_seed(config.SEED)
cp.random.seed(config.SEED)

print(f"Tensorflow ver. {tf.__version__}")

# verify GPU devices are available and ready
os.environ['CUDA_VISIBLE_DEVICES'] = config.CUDA
devices = tf.config.list_physical_devices('GPU')
assert len(devices) != 0, "No GPU devices found."

# ------------------------------------------------------------------
# System Configurations
# ------------------------------------------------------------------
if config.MIXED_PRECISION:
    policy = mixed_precision.Policy('mixed_float16')
    mixed_precision.set_global_policy(policy)
    print('Mixed precision enabled')

if config.XLA_ACCELERATE:
    tf.config.optimizer.set_jit(True)
    print('Accelerated Linear Algebra enabled')

# set memory growth to infinite to account for multiple devices
physical_devices = tf.config.experimental.list_physical_devices('GPU')
for pd in physical_devices:
    configTF = tf.config.experimental.set_memory_growth(pd, True)


# ---------------------------------------------------------------------------
# script predict.py
# ---------------------------------------------------------------------------
def predict(x_data, model, config, spline, normalize=True, standardize=True):

    # open rasters and get both data and coordinates
    rast_shape = x_data[:, :, 0].shape  # shape of the wider scene

    # in memory sliding window predictions
    wsx, wsy = config.PRED_WINDOW_SIZE[0], config.PRED_WINDOW_SIZE[1]

    # if the window size is bigger than the image, predict full image
    if wsx > rast_shape[0]:
        wsx = rast_shape[0]
    if wsy > rast_shape[1]:
        wsy = rast_shape[1]

    prediction = np.zeros(rast_shape)  # crop out the window
    print(f'wsize: {wsx}x{wsy}. Prediction shape: {prediction.shape}')

    for sx in tqdm(range(0, rast_shape[0], wsx)):  # iterate over x-axis
        for sy in range(0, rast_shape[1], wsy):  # iterate over y-axis
            x0, x1, y0, y1 = sx, sx + wsx, sy, sy + wsy  # assign window
            if x1 > rast_shape[0]:  # if selected x exceeds boundary
                x1 = rast_shape[0]  # assign boundary to x-window
            if y1 > rast_shape[1]:  # if selected y exceeds boundary
                y1 = rast_shape[1]  # assign boundary to y-window
            if x1 - x0 < config.TILE_SIZE:  # if x is smaller than tsize
                x0 = x1 - config.TILE_SIZE  # assign boundary to -tsize
            if y1 - y0 < config.TILE_SIZE:  # if selected y is small than tsize
                y0 = y1 - config.TILE_SIZE  # assign boundary to -tsize

            window = x_data[x0:x1, y0:y1, :].values  # get window
            window = cp.asarray(window)

            window[window < 0] = 0  # remove lower bound values
            window[window > 10000] = 10000  # remove higher bound values

            # adding indices
            # window = cp.transpose(window, (2, 0, 1))
            # fdi = indices.fdi(
            #    window, config.PRED_BANDS_INPUT,
            #    factor=config.INDICES_FACTOR, vtype='int16'
            # )
            # si = indices.si(
            #    window, config.PRED_BANDS_INPUT,
            #    factor=config.INDICES_FACTOR, vtype='int16'
            # )
            # ndwi = indices.ndwi(
            #    window, config.PRED_BANDS_INPUT,
            #    factor=config.INDICES_FACTOR, vtype='int16'
            # )
            # print(fdi.shape, si.shape, ndwi.shape, window.shape)

            # concatenate all indices
            # window = cp.concatenate((window, fdi, si, ndwi), axis=0)
            # window = cp.transpose(window, (1, 2, 0))

            if config.NORMALIZE:
                window = window / config.normalization_factor

            window = cp.asnumpy(window)
            print("Window shape", window.shape)

            # perform sliding window prediction
            prediction[x0:x1, y0:y1] = \
                predict_all(window, model, config, spline=spline)

    return prediction


def main():

    # Main function to collect configuration file and run the script
    # print(f'GPU REPLICAS: {strategy.num_replicas_in_sync}')

    # Generate 2-dimensional spline to avoid boundary problems
    spline = _2d_spline(config.TILE_SIZE, power=2)

    # Loading the trained model
    model = load_model(config.MODEL_NAME)
    model.summary()  # print summary of the model

    # Get list of files to predict
    print(f'Number of files to predict: {len(config.PRED_FILENAMES)}')

    # Tterate over files and predict them
    for fname in config.PRED_FILENAMES:

        # measure execution time
        start_time = time.perf_counter()

        # path + name to store prediction into
        save_image = \
            config.PRED_SAVE_DIR + fname[:-4].split('/')[-1] + '_pred.tif'

        # --------------------------------------------------------------------------------
        # if prediction is not on directory, start predicting
        # (allows for restarting script if it was interrupted at some point)
        # --------------------------------------------------------------------------------
        if not os.path.isfile(save_image):

            print(f'Starting to predict {fname}')

            # --------------------------------------------------------------------------------
            # Extracting and resizing test and validation data
            # --------------------------------------------------------------------------------
            x_data = xr.open_rasterio(fname, chunks=config.DASK_SIZE)
            x_data = x_data.transpose("y", "x", "band")
            x_data = x_data[:, :, :4]

            # --------------------------------------------------------------------------------
            # Calculate missing indices
            # --------------------------------------------------------------------------------

            # --------------------------------------------------------------------------------
            # Getting predicted labels
            # --------------------------------------------------------------------------------
            os.system('mkdir -p {}'.format(config.PRED_SAVE_DIR))
            prediction = predict(
                x_data, model, config, spline, normalize=config.NORMALIZE,
                standardize=config.STANDARDIZE
            )
            prediction = prediction.astype(np.int8)  # type to int16

            # --------------------------------------------------------------------------------
            # Generating visualization from prediction
            # --------------------------------------------------------------------------------
            arr_to_tif(raster_f=fname, segments=prediction, out_tif=save_image)
            del prediction

        # This is the case where the prediction was already saved
        else:
            print(f'{save_image} already predicted.')

        print(f'Time: {(time.perf_counter() - start_time)}')


# -------------------------------------------------------------------------------
# main
# -------------------------------------------------------------------------------

if __name__ == "__main__":

    main()


scripts/postprocessing.py
import os
import glob
import cv2
from tqdm import tqdm
import xarray as xr
import numpy as np  # for arrays modifications
from skimage import color
import rasterio.features as riofeat  # rasterio features include sieve filter
import rasterio as rio
from scipy.ndimage import median_filter, binary_fill_holes

__author__ = "Jordan A Caraballo-Vega, Science Data Processing Branch"
__email__ = "jordan.a.caraballo-vega@nasa.gov"
__status__ = "Development"


def npy_to_tif(raster_f='image.tif', segments='segment.npy',
               outtif='segment.tif', ndval=-9999
               ):
    """
    Args:
        raster_f:
        segments:
        outtif:
    Returns:
    """
    # get geospatial profile, will apply for output file
    with rio.open(raster_f) as src:
        meta = src.profile
        nodatavals = src.read_masks(1).astype('int16')
    print(meta)

    # load numpy array if file is given
    if type(segments) == str:
        segments = np.load(segments)
    segments = segments.astype('int16')
    print(segments.dtype)  # check datatype

    nodatavals[nodatavals == 0] = ndval
    segments[nodatavals == ndval] = nodatavals[nodatavals == ndval]

    out_meta = meta  # modify profile based on numpy array
    out_meta['count'] = 1  # output is single band
    out_meta['dtype'] = 'int16'  # data type is float64

    # write to a raster
    with rio.open(outtif, 'w', **out_meta) as dst:
        dst.write(segments, 1)


def scale_contour(cnt, scale):
    M = cv2.moments(cnt)
    try:
        cx = int(M['m10']/M['m00'])
        cy = int(M['m01']/M['m00'])
    except ZeroDivisionError:
        cx, cy = 0, 0

    cnt_norm = cnt - [cx, cy]
    cnt_scaled = cnt_norm * scale
    cnt_scaled = cnt_scaled + [cx, cy]
    cnt_scaled = cnt_scaled.astype(np.int32)

    return cnt_scaled


def main():

    # CNN
    # root_dir = '/att/gpfsfs/atrepo01/ILAB/projects/Vietnam/Jordan/vietnam_cm_models/predictions/100_unet_viet_cm_4chann_std_Adadelta_256_0.0001_128_1000.h5'
    # out_dir = f'{root_dir}/processed'
    
    # RF
    #root_dir = '/att/gpfsfs/atrepo01/ILAB/projects/Vietnam/Jordan/Vietnam_CLOUD/cloud_predictions_rf/cloudmask_keelin_4bands_rf'
    #out_dir = f'{root_dir}/processed_rf'

    root_dir = '/Users/jacaraba/Documents/Development/ILAB/ai-cloud-shadow-masking-vhr/projects/cnn_cloud_v3/data/alaska/predictions'
    out_dir = '/Users/jacaraba/Documents/Development/ILAB/ai-cloud-shadow-masking-vhr/projects/cnn_cloud_v3/data/alaska/predictions/post'

    os.system('mkdir -p {}'.format(out_dir))

    #mask_filenames = f'{root_dir}/images/*.tif'
    mask_filenames = f'{root_dir}/*.tif'
    mask_filenames = glob.glob(mask_filenames)

    for mask_f in tqdm(mask_filenames):

        # open mask
        raster_mask = xr.open_rasterio(mask_f, chunks={'band': 1, 'x': 2048, 'y': 2048})

        # out filename
        out_filename = out_dir + '/' + mask_f.split('/')[-1]

        output = np.squeeze(raster_mask.values)  # values

        # riofeat.sieve(prediction, size, out, mask, connectivity)
        riofeat.sieve(output, 800, output, None, 8)

        # median filter and binary filter
        output = median_filter(output, size=20)
        output = binary_fill_holes(output).astype(int)

        # increase size of contour
        output = np.uint8(np.squeeze(output) * 255)
        ret, thresh = cv2.threshold(output, 127, 255, 0)

        contours, hierarchy = cv2.findContours(
            thresh, cv2.RETR_TREE, cv2.CHAIN_APPROX_NONE
        )

        conts = np.zeros((thresh.shape[0], thresh.shape[1], 3))

        for c in contours:
            c = scale_contour(c, 1.12)  # 1.05 looks decent
            conts = cv2.drawContours(conts, [c], -1, (0, 255, 0), cv2.FILLED)

        # converting mask into raster
        conts = color.rgb2gray(conts).astype('int16')
        conts[conts > 0] = 1

        npy_to_tif(raster_f=mask_f, segments=conts, outtif=out_filename)


# -------------------------------------------------------------------------------
# main
# -------------------------------------------------------------------------------

if __name__ == "__main__":

    main()


scripts/config/Config.py
# --------------------------------------------------------------------------
# Configuration of the parameters for training and preprocessing
# NGA dataset using deep learning techniques.
# --------------------------------------------------------------------------
import glob              # get global files from directory
import tensorflow as tf  # deep learning framework

__author__ = "Jordan A Caraballo-Vega, Science Data Processing Branch"
__email__ = "jordan.a.caraballo-vega@nasa.gov"
__status__ = "Development"
__date__ = "02/10/2021"


class Configuration:

    def __init__(self):

        # ------------------------------------------------------------------
        # System configurations
        # ------------------------------------------------------------------
        # self.SEED: define static random seed value reproducibility
        self.SEED = 42

        # self.CUDA: define CUDA devices to use
        self.CUDA = "0,1,2,3"

        # self.FOLDS: define folds for batch iterations
        self.FOLDS = 3

        # ------------------------------------------------------------------
        # Mixed Precision, XLA, Multi-GPU strategies
        # ------------------------------------------------------------------
        # self.MIRROR_STRATEGY: multi-gpu strategy
        self.MIRROR_STRATEGY = True

        # self.MIXED_PRECISION: TF mixed precision enable
        self.MIXED_PRECISION = True

        # self.XLA_ACCELERATE: XLA acceleration enabled
        self.XLA_ACCELERATE = False

        # self.AUTOTUNE: tensorflow autotuning feature enabled
        # For more information about autotune:
        # https://www.tensorflow.org/guide/data_performance#prefetching
        self.AUTOTUNE = tf.data.experimental.AUTOTUNE

        # ------------------------------------------------------------------
        # General imagery variables
        # ------------------------------------------------------------------
        # self.NODATA_VAL: no data value present on the imagery
        self.NODATA_VAL = 9999

        # self.DASK_SIZE: directory for data structure out of memory mapping
        self.DASK_SIZE = {'band': 1, 'x': 2048, 'y': 2048}

        # self.INIT_BANDS: bands available in the original data
        self.INIT_BANDS = \
            ['Blue', 'Green', 'Red', 'NIR1', 'HOM1', 'HOM2']

        # self.N_CLASSES: number of classes that will be classified
        # An example of this: tree, shadow, cloud, background = 4 classes
        self.N_CLASSES = 2

        # self.TILE_SIZE: tile size to feed the model
        self.TILE_SIZE = 256

        # ------------------------------------------------------------------
        # Preprocessing
        # ------------------------------------------------------------------
        # self.PREP_DATA_FILE: includes the raster filenames to use for train
        # Look at the README for more details on what this file includes.
        self.PREP_DATA_FILE = 'config/data.csv'

        # self.PREP_DATA_INPDIR: directory where GeoTIF data rasters reside
        self.PREP_DATA_INPDIR = '/att/projects/Vietnam/data'

        # self.PREP_LABELS_INPDIR: directory where GeoTIF label rasters reside
        self.PREP_LABELS_INPDIR = '/att/projects/Vietnam/labels'

        # self.PREP_BANDS_OUTPUT: bands to output from data to train
        self.PREP_BANDS_OUTPUT = \
            ['Blue', 'Green', 'Red', 'NIR1']

        # self.PREP_ROOT_OUTDIR: root directory that leads to data subdirs
        self.PREP_ROOT_OUTDIR = '/att/projects/Vietnam/dataset'


        self.PREP_DATA_INPDIR = '/att/gpfsfs/atrepo01/ILAB/projects/Vietnam/Sarah/data'
        self.PREP_LABELS_INPDIR = '/att/gpfsfs/atrepo01/ILAB/projects/Vietnam/Mary/cloud_labels'
        self.PREP_ROOT_OUTDIR = '/att/gpfsfs/atrepo01/ILAB/projects/Vietnam/Jordan/test_container'




        # self.PREP_TRAIN_OUTDIR: output directory to store training data
        self.PREP_TRAIN_OUTDIR = self.PREP_ROOT_OUTDIR + '/train'

        # self.PREP_VAL_OUTDIR: output directory to store validation data
        self.PREP_VAL_OUTDIR = self.PREP_ROOT_OUTDIR + '/val'

        # self.NORMALIZE: bool value to apply normalization to the dataset
        self.NORMALIZE = False

        # self.normalization_factor: float value to accomply normalization
        self.normalization_factor = 65535.0

        # self.STANDARDIZE: bool value to apply standardization to the dataset
        self.STANDARDIZE = True

        # self.INDICES_FACTOR: float value to calculate additional bands
        # related to TOA values generated from EVHR
        self.INDICES_FACTOR = 10000.0

        # ------------------------------------------------------------------
        # Training
        # ------------------------------------------------------------------
        # self.TRAIN_BANDS_INPUT: bands that will be sent as input for train
        self.TRAIN_BANDS_INPUT = self.PREP_BANDS_OUTPUT

        # Defining model variables

        # self.N_CHANNELS: number of channels (number of bands) to train on
        self.N_CHANNELS = len(self.TRAIN_BANDS_INPUT)

        # self.INPUT_SIZE: input size to feed neural network
        self.INPUT_SIZE = (self.TILE_SIZE, self.TILE_SIZE, self.N_CHANNELS)

        # self.START_EPOCH: epoch to start from training
        self.START_EPOCH = 0

        # self.N_EPOCHS: number of epochs to run for
        self.N_EPOCHS = 1000

        # self.BATCH_SIZE: batch size to feed the network
        # 128 bsize works well with 32GB RAM GPU
        self.BATCH_SIZE = 128
        self.VAL_BATCH_SIZE = 512

        # self.TRAIN_ROOT_INPDIR: root dir where train and val data resides
        self.TRAIN_ROOT_INPDIR = self.PREP_ROOT_OUTDIR

        # self.TRAIN_DATADIR: directory where train data resides
        self.TRAIN_DATADIR = self.PREP_TRAIN_OUTDIR

        # self.VAL_DATADIR: directory where val data resides
        self.VAL_DATADIR = self.PREP_VAL_OUTDIR

        # Training hyper-parameters
        self.MODEL_METADATA = {
            'id': '100',
            'network': 'unet',
            'network_maps': [64, 128, 256, 512, 1024],
            'lr': 0.0001,
            'momentum': 0.90,
            'gradient': 0.95,
            'loss': 'categorical_crossentropy',
            'optimizer_name': 'Adadelta',
            'metrics': ['accuracy'],
            'do_aug': True
        }

        # Model description and output directory

        # self.MODEL_NAME: model name to train
        self.MODEL_NAME = '100_unet_viet_cm_4chann_std_Adadelta_256_0.0001_128'

        # self.MODEL_SAVEDIR: directory to save trained models
        self.MODEL_SAVEDIR = f'/att/projects/Vietnam/unet/{self.MODEL_NAME}'



        self.MODEL_SAVEDIR = f'{self.PREP_ROOT_OUTDIR}/{self.MODEL_NAME}'



        # self.MODEL_OUTPUT_NAME: output name for model to be saved
        self.MODEL_OUTPUT_NAME = f'{self.MODEL_SAVEDIR}/{self.MODEL_NAME}.h5'

        # Training extensions - Callbacks
        self.CALLBACKS = \
            ['ModelCheckpoint', 'CSVLogger', 'EarlyStopping', 'TensorBoard']

        self.CALLBACKS_METADATA = {
            'patience_earlystop': 20,
            'patience_plateu': 5,
            'monitor_earlystop': 'val_loss',
            'monitor_checkpoint': 'val_loss',
            'save_freq': 'epoch',
            'factor_plateu': 0.20,
            'min_lr_plateu': 0.00,
            'history_freq': 5,
            'save_best_only': True
        }

        # ------------------------------------------------------------------
        # Prediction
        # ------------------------------------------------------------------
        # self.PRED_BANDS_INPUT: bands available in the original data
        self.PRED_BANDS_INPUT = self.INIT_BANDS

        # self.PRED_BANDS_OUTPUT: number of bands to feed model
        # if the number of bands is greater, indices need to be calculated
        self.PRED_BANDS_OUTPUT = self.PREP_BANDS_OUTPUT

        # self.PRED_BSIZE: batch size for concurrent imagery predictions
        self.PRED_BATCH_SIZE = 128

        # self.PRED_FILENAMES: regex to flag all files to predict
        self.PRED_FILENAMES = '/att/projects/Vietnam/images/*.tif'


        self.PRED_FILENAMES = '/att/gpfsfs/atrepo01/ILAB/projects/Vietnam/Jordan/data_keelin/*.tif'

        
        self.PRED_FILENAMES = glob.glob(self.PRED_FILENAMES)

        # self.PRED_OVERLAP: overlap between tiles to predict
        self.PRED_OVERLAP = 0.25

        # self.PRED_WINDOW_SIZE: greatest window size that can fit into memory
        self.PRED_WINDOW_SIZE = [8192, 8192]

        # self.MODEL_STR: specific model to open based on epoch
        self.MODEL_STR = f"{self.MODEL_NAME}_66.h5"

        # self.MODEL_NAME: model path and model name to open
        self.MODEL_NAME = f'{self.MODEL_SAVEDIR}/{self.MODEL_STR}'

        # self.PRED_SAVE_DIR: directory to save GeoTIF predictions on
        self.PRED_SAVE_DIR = \
            f'{self.PREP_ROOT_OUTDIR}/predictions/{self.MODEL_STR}/images'

        # self.PRED_SEG_SAVE_DIR: directory to save numpy predictions on
        self.PRED_SEG_SAVE_DIR = \
            f'{self.PREP_ROOT_OUTDIR}/predictions/{self.MODEL_STR}/segments'
